natsstreaming/nats_streaming.go

```diff
@@ -57,80 +56,49 @@ type asyncMessageSink struct {
 
 func (p *asyncMessageSink) PublishMessages(ctx context.Context, acks chan<- substrate.Message, messages <-chan substrate.Message) (rerr error) {
 
-	var wg sync.WaitGroup
-
 	ctx, cancel := context.WithCancel(ctx)
-	defer func() {
-		cancel()
-		wg.Wait()
-	}()
+	defer cancel()
 
 	conn := p.sc
 
-	natsAcks := make(chan string)
 	natsAckErrs := make(chan error, 1)
-
-	type toAckType struct {
-		guid    string
-		message substrate.Message
-	}
-
-	toAck := make(chan toAckType)
 	ackErrs := make(chan error, 1)
+	publishErr := make(chan error, 1)
 
-	wg.Add(1)
 	go func() {
-		defer wg.Done()
-		ackMap := make(map[string]substrate.Message)
+	LOOP:
 		for {
 			select {
-			case natsGUID := <-natsAcks:
-				msg := ackMap[natsGUID]
-				if msg == nil {
-					ackErrs <- fmt.Errorf("got ack from nats streaming for unknown guid %v", natsGUID)
-					return
-				}
-				delete(ackMap, natsGUID)
-				select {
-				case acks <- msg:
-				case <-ctx.Done():
-					//return ctx.Err()
-					return
-				}
-			case ta := <-toAck:
-				ackMap[ta.guid] = ta.message
 			case <-ctx.Done():
-				return
-			}
-		}
-	}()
-
-	for {
-		select {
-		case <-ctx.Done():
-			//return ctx.Err()
-			return nil
-		case msg := <-messages:
-			guid, err := conn.PublishAsync(p.subject, msg.Data(), func(guid string, err error) {
-				if err != nil {
-					select {
-					case natsAckErrs <- err:
-					default:
+				//return ctx.Err()
+				break LOOP
```

Comment:
I think it's nicer to just `return` here instead of breaking of the named loop, but that's just my personal preference as I'm not big fan of named loops 😅

Reply:
It seemed to never return. I agree that it looks ugly, however has served me well so far, and is explicit. I also normally have some tidy up at the bottom of the for loops, so for me that's less I have to remember.